5513 database variablemetadata

lubitchv opened the pull request:
## New Contributors
This relates to issue #5513
New classes VariableMetadata, CategoryMetadata and VarGroup are added. These classes will correspond to the new tables in the database that are required for Data Curation Tool, VariableMetadata for mutable variable level metadata, CategoryMetadata for mutable category data and VarGroup for grouping variables in groups.
Universe and weighted fields are removed from datavariable, since they are empty in the present dataverse and they are moved to VariableMetadata.
Removing universe and weighted from datavariable table is added to upgrade sql script.
This is just a first step in #4448

connects to #5513

pdurbin commented on scripts/database/upgrades/upgrade_v4.10.1_to_4.11.sql:
```diff
@@ -1,2 +1,6 @@
+
+ALTER TABLE datavariable
+DROP COLUMN  if exists universe,
+DROP COLUMN  if exists weighted;
```
These two columns ("universe" and "weighted") are being dropped? Are they always empty?

Later I noticed that the pull request says "Universe and weighted fields are removed from datavariable, since they are empty in the present dataverse and they are moved to VariableMetadata." So that answers my question. A comment in the script would be nice. Not everyone will see the pull request.

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/datavariable/VariableMetadata.java:
```diff
@@ -0,0 +1,210 @@
+
+    /*
+     * fileMetadta: FileMetadata to which this metadata belongs.
+     */
```
These could be converted to Javadoc, starting with two `**` instead of one `*`.

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/datavariable/VariableServiceBean.java:
```diff
@@ -51,6 +51,15 @@ public DataVariable find(Object pk) {
+        TypedQuery<VariableMetadata> query = em.createQuery("SELECT object(o) FROM VariableMetadata as o where o.dataVariable.id =:value1 and o.fileMetadata.id =:value2", VariableMetadata.class);
+
+        query.setParameter("value1", datVarId);
+        query.setParameter("value2", metaId);
```
Is there a reason why these are called "value1" and "value2" instead of more meaningful names?

lubitchv commented:
@pdurbin Thanks for comments. I made the changes that you requested.

pdurbin commented:
@lubitchv I see you added a comment to the SQL script. Thanks!

lubitchv commented:
@pdurbin @landreev 
I made a mistake introducing wfreq (weighted frequency) as Double in VariableMetadata, wreq is related to category. Therefore I introduced new class CategoryMetadata for mutable category fields such as weighted frequency. In VariableMetadata I added Collection of CategoryMetadatas instead of wfreq.

lubitchv commented:
@landreev Thank you. Yes, it is ok with us. I will leave "weighted" in DataVariable and will update the code accordingly for this pull request.
